### PR TITLE
Add UserStorage contract

### PR DESCRIPTION
**Pull request by pavelisnear**
This PR adds a new contract that implements [StorageMangement](https://github.com/near/near-sdk-js/blob/develop/packages/near-contract-standards/src/storage_management/index.ts#L24) interface.

Fixes https://github.com/near/chain-deployed-ui/issues/14

### CODE REVIEW

**Comment by mpeterdev** [comment]
@race-of-sloths

**Comment by race-of-sloths** [comment]
### 🙁 This repository is not a part of the [Race of Sloths](https://race-of-sloths.com)
Please ask the maintainer to [contact us](https://github.com/NEAR-DevHub/race-of-sloths/issues/new/choose) to join the race!

<details><summary><strong>What is the Race of Sloths</strong></summary>
<p>

Race of Sloths is a friendly competition where you can participate in challenges and compete with other open-source contributors within your normal workflow

For contributors:

- Tag us inside your pull requests
- Wait for the maintainer to review and score your pull request
- Check out your position in the [Leaderboard](https://race-of-sloths.com/leaderboard)
- Keep weekly and monthly streaks to reach higher positions
- Boast your contributions with a dynamic picture of your [Profile](https://race-of-sloths.com/profile/pavelisnear)

For maintainers:
- Score pull requests that participate in the Race of Sloths
- Engage contributors with fair scoring and fast responses so they keep their streaks
- Promote the Race to the point where the Race starts promoting you
- Grow the community of your contributors

Feel free to check [our website](https://race-of-sloths.com) for additional details!

</p>
</details>

**Comment by mpeterdev** [comment]
@race-of-sloths

**Comment by mpeterdev** [comment]
@race-of-sloths score 5

**Comment by race-of-sloths** [comment]
### 🌟 Score recorded!

@mpeterdev, thank you for scoring this pull request in the Race of Sloths!

File: packages/paid-storage-contract/src/interface.ts
Comment (mpeterdev) [inline packages/paid-storage-contract/src/interface.ts:1]: this can probably be deleted right? I am not seeing it used

**Review by mpeterdev — CHANGES_REQUESTED** [review]

File: packages/paid-storage-contract/src/contract.ts
Comment (andy-haynes) [inline packages/paid-storage-contract/src/contract.ts:73]: Is this required if `amount` is already typed as `bigint`? Does it still come in as a string?

File: packages/paid-storage-contract/src/contract.ts
Comment (mpeterdev) [inline packages/paid-storage-contract/src/contract.ts:121]: can be `view` call

File: packages/paid-storage-contract/src/contract.ts
Comment (pavelisnear) [reply packages/paid-storage-contract/src/contract.ts:53]: yeah, that's a fair note. I'm leaving a placeholder for now to update the refunds logic.

File: packages/paid-storage-contract/src/contract.ts
Comment (pavelisnear) [reply packages/paid-storage-contract/src/contract.ts:73]: yeah, I've had an issue that before adding this line the contract was saying that it's a wrong type (I guess it was string instead). So I've found an inspiration here -> https://github.com/near/near-sdk-js/blob/develop/packages/near-contract-standards/src/fungible_token/core_impl.ts#L314 and adding this casing has resolved the issue.

File: packages/paid-storage-contract/src/contract.ts
Comment (pavelisnear) [reply packages/paid-storage-contract/src/contract.ts]: Fixed

**Review by mpeterdev — CHANGES_REQUESTED** [review]

File: packages/paid-storage-contract/src/contract.ts
Comment (pavelisnear) [reply packages/paid-storage-contract/src/contract.ts:86]: yeah, good catch!

File: packages/paid-storage-contract/src/contract.ts
Comment (pavelisnear) [reply packages/paid-storage-contract/src/contract.ts:158]: This check was implemented according to the NEP note on `storage_unregister`:
```
// If `force=false` or `force` is omitted, the contract MUST panic if caller
// has existing account data, such as a positive registered balance (eg token
// holdings).
```

I think this implementation does not contradict with the idea:
```
logic should be that if a user has any tied up storage besides the registration min (meaning they have dapp code stored) then unregister fails unless force is provided, in which case force would also wipe their data
```
since `storage besides the registration min` is exactly what `.available` value represents.

**Update!**
After taking a second look at the NEP description of the method, I think that this line `// Unregisters the predecessor account and returns the storage NEAR deposit.` contradicts with the rest of the description, since f `force=true` it should burn balances, while `force=false` or `force` is omitted, the contract MUST panic. So it seems to me that there is no way when the storage NEAR deposit is being refunded.

Comment (mpeterdev) [reply packages/paid-storage-contract/src/contract.ts:155]: This confused me at first too. The positive registered balance is referring to fungible token balance, not storage balance. If force is true, then the users data is forcefully wiped even if they have things they might care about like fungible tokens, NFTS, or in our case dapp code

Comment (pavelisnear) [reply packages/paid-storage-contract/src/contract.ts:155]: I've updated the logic, so now if `force=false` or `force` is omitted, we do a refund of the account registration costs only. I've also added a check just in case the `total` is less than the current registration cost (but it might be a redundant one).
Technically it means that the user doing a force unregister will get a refund of the storage registration cost only and the available funds (if there are any) will stay on the contract balance.

```
const min_balance: Balance = this.storage_balance_bounds().min;
const refund = storage_balance.total >= min_balance ? min_balance : storage_balance.total;
NearPromise.new(predecessor_account_id).transfer(refund);
```

@mpeterdev is the anything we could improve here in your opinion?

Comment (pavelisnear) [reply packages/paid-storage-contract/src/contract.ts:155]: **UPDATE**

`storage_unregister` with `force=true` deletes the account record and refund the total user balance.

**Review by mpeterdev — APPROVED** [review]

**Review by andy-haynes — CHANGES_REQUESTED** [review]
Just a couple missing `@call({ payableFunction: true })` otherwise LGTM 🎉

**Comment by race-of-sloths** [comment]
### ✅ PR is finalized! 
Your contribution is much appreciated with a final score of 5!
You have received 60 (50 base + 10 weekly bonus) Sloth points for this contribution


Another weekly streak completed, well done @pavelisnear! To keep your weekly streak and get another bonus make pull request next week! Looking forward to see you in race-of-sloths

**Pull request merged**
